### PR TITLE
[SPARK-9428][SQL] Add test cases for null inputs for expression unit tests

### PR DESCRIPTION
**Pull request by yjshen**
JIRA: https://issues.apache.org/jira/browse/SPARK-9428

### CODE REVIEW

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala [hunk 1]
```diff
@@ -43,6 +43,45 @@ class CastSuite extends SparkFunSuite with ExpressionEvalHelper {
     checkEvaluation(cast(v, Literal(expected).dataType), expected)
   }
 
+  private def checkNullCast(from: DataType, to: DataType): Unit = {
+    checkEvaluation(Cast(Literal.create(null, from), to), null)
+  }
+
+  test("null cast") {
+    val numericTypes = Seq(IntegerType, ShortType, ByteType, LongType,
```
Comment (JoshRosen) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala]: I found myself writing these lists a lot in my own tests, so I defined some similar ones as constants in `DataTypeTestUtils`.

Comment (yjshen) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala]: Thanks, I will use that in the next commit.

**Comment by rxin** [comment]
cc @davies

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala [hunk 2]
```diff
@@ -1035,7 +1034,11 @@ case class RegExpReplace(subject: Expression, regexp: Expression, rep: Expressio
             m.appendTail(${termResult});
             ${ev.primitive} = ${classNameUTF8String}.fromString(${termResult}.toString());
             ${ev.isNull} = false;
+          } else {
+            ${ev.isNull} = true;
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala]: use `true` as the initial value.

**Comment by davies** [comment]
LGTM, just one minor comment, thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala (hunk 2 above)
Comment (yjshen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala]: I found `UnaryExpression` and `BinaryExpression` both use `${eval.isNull}` as the initial value of `${ev.isNull}`, therefore I did the same here.
If it's not a convention I would just use `true` as initial value and remove these two `else { ${ev.isNull} = true;` here.
And do the same thing in `RegExpExtract` codegen as well

**Comment by rxin** [comment]
Note that this no longer merges cleanly.

**Comment by yjshen** [comment]
yes, I've found this and rebasing the code now.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala (hunk 2 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala]: For UnaryExpression, it's better to use `ev.isNull = eval.isNull`, for others, using `ev.isNull = true` will be better.

Comment (yjshen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala]: make sense, will update this.

**Comment by yjshen** [comment]
Jenkins, retest this please.

**Comment by yjshen** [comment]
@davies, I've added null input tests in MathFunctionSuite, and update codegen in RegExpExtract and RegExpReplace. can you review this again? Thanks!

**Comment by davies** [comment]
LGTM, waiting for tests.

**Comment by davies** [comment]
Merging this into master, thanks!

**Pull request closed without merging**